### DISCUSSION TITLE
UART3 (PB10/PB11) on STM32F103xB

### DISCUSSION THREAD

**Discussion by macbef** [grblHAL/core#585] [General]
Hello,
Is there a way to use the native UART3 (on pins PB10/PB11) on an STM32F103RC instead of USB?

Just setting USB_SERIAL_CDC=0 doesn't help (maybe this is enabling UART1?)

Any help would be highly appreciated!

**Comment by terjeio** [comment]
Should be possible, depending on the board map. Which one?

**Comment by macbef** [comment]
I've built a custom board with an opto-isolated RS232 on pins PB10(tx) and PB11(rx) on an STM32F103RC.

board = genericSTM32F103RC

I have a my_machine_map that works great with USB.
Now I started to play around with the RS232 but I can't get it to work (not even on UART1)

Do I need flow control pins as well? As for now I'm only using TX and RX.

**Reply by terjeio** [reply]
Adding `#define SERIAL_PORT  3` in the map file should work.

> Do I need flow control pins as well?

Not unless you need it. But you may have to wire and/or configure the port on the PC side accordingly.

**Reply by macbef** [reply]
Are you sure with the `#define SERIAL_PORT 3`?
I tried to find the the string SERIAL_PORT in the project directory and couldn't find a single reference.

> Not unless you need it. But you may have to wire and/or configure the port on the PC side accordingly.

I tried to compile for UART1 (with USB disabled) and use my UART adapter that I use for programming/flashing. So I know my UART communication should be working. On the PC side I'm not using CTS. But even here I get no communication.

**Reply by macbef** [reply]
Or do I need SERIAL_MOD?

**Reply by terjeio** [reply]
> I tried to find the the string SERIAL_PORT in the project directory and couldn't find a single reference.

You have an older version of the driver? The later ones has it referenced [here](https://github.com/grblHAL/STM32F1xx/blob/361dc16fa677ef7149400adb406ea2a70a7ab0f5/Src/serial.c#L77-L87).

**Reply by macbef** [reply]
... yes, I cloned the repo in January 2023 and never dared to update ... never change a running system :)
The serial.c file looks very different. I'll update to the latest commit and give it a try. Thanks so far!!

**Comment by macbef** [comment]
Nice! With the current repo and

`#define SERIAL_PORT 3`

it works out of the box!

Thanks for your support!
